Sources/Classes/RSContext.m

```diff
@@ -71,119 +85,159 @@ - (void) createAndPersistTraits {
+                        [externalIds removeObjectAtIndex:index];
+                        index--;
+                        break;
+                    }
+                }
+            }
+        }
+
+        if ([externalIds count]) {
+            [_externalIds addObjectsFromArray: externalIds];
+        }
+    });
 }
 
 - (void)persistExternalIds {
-    if (_externalIds != nil) {
-        // update persistence storage
-        NSData *externalIdJsonData = [NSJSONSerialization dataWithJSONObject:[RSUtils serializeArray:[_externalIds copy]] options:0 error:nil];
-        NSString *externalIdJson = [[NSString alloc] initWithData:externalIdJsonData encoding:NSUTF8StringEncoding];
-        
-        [preferenceManager saveExternalIds:externalIdJson];
-    }
+    dispatch_async(queue, ^{
+        if (_externalIds != nil) {
+            // update persistence storage
+            NSData *externalIdJsonData = [NSJSONSerialization dataWithJSONObject:[RSUtils serializeArray:[_externalIds copy]] options:0 error:nil];
+            NSString *externalIdJson = [[NSString alloc] initWithData:externalIdJsonData encoding:NSUTF8StringEncoding];
+            [preferenceManager saveExternalIds:externalIdJson];
+        }
+    });
 }
 
 - (void)resetExternalIds {
-    _externalIds = nil;
-    [preferenceManager clearExternalIds];
+    dispatch_async(queue, ^{
+        _externalIds = nil;
+        [preferenceManager clearExternalIds];
+    });
 }
 
 - (void)putAppTrackingConsent:(int)att {
-    if (att < RSATTNotDetermined) {
-        _device.attTrackingStatus = RSATTNotDetermined;
-    } else if (att > RSATTAuthorize) {
-        _device.attTrackingStatus = RSATTAuthorize;
-    } else {
-        _device.attTrackingStatus = att;
-    }
+    dispatch_async(queue, ^{
+        if (att < RSATTNotDetermined) {
+            _device.attTrackingStatus = RSATTNotDetermined;
+        } else if (att > RSATTAuthorize) {
+            _device.attTrackingStatus = RSATTAuthorize;
+        } else {
+            _device.attTrackingStatus = att;
+        }
+    });
 }
 
 - (NSDictionary<NSString *,NSObject *> *)dict {
-    NSMutableDictionary *tempDict = [[NSMutableDictionary alloc] init];
-    [tempDict setObject:[_app dict] forKey:@"app"];
-    [tempDict setObject:[RSUtils serializeDict:_traits] forKey:@"traits"];
-    [tempDict setObject:[_library dict] forKey:@"library"];
-    [tempDict setObject:[_os dict] forKey:@"os"];
-    [tempDict setObject:[_screen dict] forKey:@"screen"];
-    if (_userAgent) {
-        [tempDict setObject:_userAgent forKey:@"userAgent"];
-    }
-    
-    [tempDict setObject:_locale forKey:@"locale"];
-    [tempDict setObject:[_device dict] forKey:@"device"];
-    [tempDict setObject:[_network dict] forKey:@"network"];
-    [tempDict setObject:_timezone forKey:@"timezone"];
-    if (_externalIds != nil) {
-        [tempDict setObject:_externalIds forKey:@"externalId"];
+    NSMutableDictionary *tempDict;
+    @synchronized (tempDict) {
```

Comment:
Can you explain why you are doing this?

Reply:
Agree `Synchronised` block is not needed since `dispatch_sync` is already there. I'll improve this.
Why I implemented `Synchronised` because if in future this `dict` method get called from multiple places then `tempDict` could be initialised multiple times and `dispatch_sync` is implemented because (the block of code which is inside it) needs to be synchronised.